### PR TITLE
Capture exceptions thrown during compilation

### LINKED ISSUE #3870: Re-requesting syntax errored view results in request hang.
Repro:

1. Open MvcSandBox's `Views/Home/Index.cshtml`
1. Add a syntax errored piece of C#: `@DateTime.Nw`
1. Run application
1. Hit page (You get error in console window)
1. Hit page a second time, browser hangs indefinitely.

### PR DESCRIPTION
**Pull request by pranavkm**
Fixes #3870

### CODE REVIEW

**Comment by pranavkm** [comment]
cc @NTaylorMullen

File: test/Microsoft.AspNet.Mvc.Razor.Test/Compilation/CompilerCacheTest.cs [hunk 1]
```diff
@@ -443,6 +444,67 @@ public void ConstructorNormalizesPrecompiledViewPath(string viewPath)
...
+        [Fact]
+        public void GetOrAdd_CachesCompilationExceptions()
+        {
+            // Arrange
+            var fileProvider = new TestFileProvider();
+            fileProvider.AddFile(ViewPath, "some content");
+            var cache = new CompilerCache(fileProvider);
+
+            // Act and Assert - 1
+            Assert.Throws<InvalidTimeZoneException>(() =>
```
Comment (NTaylorMullen) [inline test/Microsoft.AspNet.Mvc.Razor.Test/Compilation/CompilerCacheTest.cs]: This looks to be more apart of the Arrange section. `Assert.Throws<>(throw new Exception)` seems odd to assert.

File: test/Microsoft.AspNet.Mvc.Razor.Test/Compilation/CompilerCacheTest.cs [hunk 2]
```diff
@@ -443,6 +444,67 @@ public void ConstructorNormalizesPrecompiledViewPath(string viewPath)
...
+        [Fact]
+        public void GetOrAdd_ReturnsSuccessfulCompilationResultIfTriggerExpires()
+        {
+            // Arrange
+            var fileProvider = new TestFileProvider();
+            fileProvider.AddFile(ViewPath, "some content");
+            var cache = new CompilerCache(fileProvider);
+
+            // Act and Assert - 1
+            Assert.Throws<InvalidTimeZoneException>(() =>
```
Comment (NTaylorMullen) [inline test/Microsoft.AspNet.Mvc.Razor.Test/Compilation/CompilerCacheTest.cs]: ditto

**Comment by NTaylorMullen** [comment]
:shipit:

**Pull request closed without merging**